src/engine/Server.cpp

```diff
@@ -492,6 +494,139 @@ Awaitable<void> Server::process(
+  // The GraphUpdate consist of a template of `SparqlTripleSimpleWithGraph` and
+  // in some cases also an GraphPattern. The template contains
+  // `TripleComponent`s (s, p and o) and a `Graph` (graph). First transform
+  // everything except Variables in the template to IDs.
+  using IdOrVariable = std::variant<Id, Variable>;
+  auto transformSparqlTripleComponent =
+      [&vocab, &localVocab](TripleComponent component) -> IdOrVariable {
+    if (component.isVariable()) {
+      return std::move(component.getVariable());
+    } else {
+      return std::move(component).toValueId(vocab, localVocab);
+    }
+  };
+  auto transformGraph =
+      [&vocab,
+       &localVocab](SparqlTripleSimpleWithGraph::Graph graph) -> IdOrVariable {
+    return std::visit(
+        ad_utility::OverloadCallOperator{
+            [](const std::monostate&) -> IdOrVariable {
+              AD_CORRECTNESS_CHECK(
+                  qlever::specialIds().contains(DEFAULT_GRAPH_IRI));
+              return qlever::specialIds().at(DEFAULT_GRAPH_IRI);
+            },
+            [&vocab, &localVocab](const Iri& iri) -> IdOrVariable {
+              ad_utility::triple_component::Iri i =
+                  ad_utility::triple_component::Iri::fromIriref(iri.iri());
+              return TripleComponent(i).toValueId(vocab, localVocab);
+            },
+            [](const Variable& var) -> IdOrVariable { return var; }},
+        graph);
+  };
+  auto transformSparqlTripleSimple =
+      [&transformSparqlTripleComponent,
+       &transformGraph](SparqlTripleSimpleWithGraph triple) {
+        return std::array{transformSparqlTripleComponent(std::move(triple.s_)),
+                          transformSparqlTripleComponent(std::move(triple.p_)),
+                          transformSparqlTripleComponent(std::move(triple.o_)),
+                          transformGraph(std::move(triple.g_))};
+      };
+  std::vector<std::array<IdOrVariable, 4>> toInsertTemplates =
+      ad_utility::transform(std::move(graphUpdate.toInsert_),
+                            transformSparqlTripleSimple);
+  std::vector<std::array<IdOrVariable, 4>> toDeleteTemplates =
+      ad_utility::transform(std::move(graphUpdate.toDelete_),
+                            transformSparqlTripleSimple);
+
+  // To calculate the set of quads for the update, for each row in the template
+  // resolve the variables using each row of the result to obtain a quad without
+  // variables. These quads can then be used with `DeltaTriples`.
+  auto resolveVariable = [](const IdTable& idTable, const size_t& row,
+                            const VariableToColumnMap& variableColumns,
+                            IdOrVariable idOrVar) -> std::optional<Id> {
+    if (std::holds_alternative<Variable>(idOrVar)) {
+      auto var = std::get<Variable>(std::move(idOrVar));
+      if (!variableColumns.contains(var)) {
+        return std::nullopt;
+      } else {
+        return idTable(row, variableColumns.at(var).columnIndex_);
```

Comment:
There are some performance gains possible here (the lookup into the variable column map is basically always the same), 
but only for now add a TODO + factor the resolving out, we can then optimize it later.

Or, as I am thinking:
Don't refactor your template into variant of `Id, Variable`, but into variants of <Id, ColumnIndexOfTheVariable>`, 
that can be done rather easily and locally here.